content/blog/hard-grids-and-baselines-android-design-fidelity/index.md

```diff
@@ -0,0 +1,289 @@
+
+![A subtitle block showing 56sp height despite the text visually being much shorter](./images/56sp.png)
+
+This means that designers, alongside developers, can force the bounds of a TextView to match the design specs and open the door to perfect implementations of their mockups.
+
+This is something I’ve personally tested in an app I designed. [**Memoire**, a note taking app](http://tiny.cc/getmemoire) for Android, is a 1:1 recreation of its mockups — for every single screen. This was made possible due to these APIs — *and because [**@sasikanth**](https://twitter.com/its\_sasikanth) is not confrontational* — , since text is what almost always makes baseline alignment and hard grids impossible to implement in production.
+
+![Near-perfect duplication of guidelines against Memoire's mockups and actual app](./images/Memoire_Bounds_and_Baselines.gif)
+
+*Memoire’s TextViews are all customized using these APIs.*
+
+## What is the purpose of firstBaselineToTopHeight and lastBaselineToBottomHeight?
+
+In reality, the new attributes were actually made to be used when creating layouts: you want to make sure the baseline is a certain distance from another element, and it also helps to align the first and lastBaseline to a 4dp grid. This mirrors the way iOS layouts are built.
+
+![A showcase of "firstBaselineToTopHeight" being used to create top-padding from an image and lower text on a card, "lastBaselineToBottomHeight" to create bottom padding against the card edge, and "lineHeight" to set the text spacing](./images/Intended_Use.png "A showcase of the various props to size this card")
+
+**However, there’s one giant flaw: You can’t align a TextView’s firstBaseline to another TextView’s lastBaseline.** So a problem immediately arises due to this limitation:
+
+<blockquote class="bigBlock"><i>What if there’s more than one TextView?</i></blockquote>
+
+As you might imagine, **if we want to keep our text aligned to a baseline grid, we need to ensure that the height of each TextView is a multiple of 4 while doing so.** This means we must apply first and lastBaseline attributes to both / all of the stacked TextViews — and that becomes hard to maintain.
+
+![A comparison table of Dos and Donts that matches the below table](./images/Dos_Donts.png)
+
+|✅ Good|🛑 Bad|
+|--|--|
+|Applying firstBaseline and lastBaseline in styles allows you to know exactly what the distance between baselines is, without having to set them one by one to ensure they properly align to a 4dp grid. | Without applying first and lastBaseline in styles, you can’t detect what the default values are, so you are forced to apply these one by one to every TextView to ensure they align to a 4dp grid. |
+
+![A comparison of how text spacing is applied on iOS and Android](./images/iOS_vs_Android.gif)
+
+The solution is to apply them in your **styles.xml **so that, when themed, the TextView is given the right text size, height, font and baseline properties.
```

Comment:
- Oxford comma
```suggestion
The solution is to apply them in your **styles.xml **so that, when themed, the TextView is given the right text size, height, font, and baseline properties.
```

Reply:
Combining with the prior suggestion:
```suggestion
The solution is to apply them in your `styles.xml` so that, when themed, the TextView is given the right text size, height, font, and baseline properties.
```